File: src/java/org/apache/fop/pdf/PDFDictionary.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* $Id$ */

package org.apache.fop.pdf;

import java.io.IOException;
import java.io.OutputStream;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.commons.io.output.CountingOutputStream;

/**
 * Class representing a PDF dictionary object
 */
public class PDFDictionary extends PDFObject {
    private boolean visited;
    /**
     * the entry map
     */
    protected Map<String, Object> entries = new java.util.HashMap<String, Object>();

    /**
     * maintains the order of the entries added to the entry map. Whenever you modify
     * "entries", always make sure you adjust this list accordingly.
     */
    protected List<String> order = new java.util.ArrayList<String>();

    /**
     * Create a new dictionary object.
     */
    public PDFDictionary() {
        super();
    }

    /**
     * Create a new dictionary object.
     * @param parent the object's parent if any
     */
    public PDFDictionary(PDFObject parent) {
        super(parent);
    }

    /**
     * Puts a new name/value pair.
     * @param name the name
     * @param value the value
     */
    public void put(String name, Object value) {
        if (value instanceof PDFObject) {
            PDFObject pdfObj = (PDFObject)value;
            if (!pdfObj.hasObjectNumber()) {
                pdfObj.setParent(this);
            }
        }
        if (!entries.containsKey(name)) {
            this.order.add(name);
        }
        this.entries.put(name, value);
    }

    /**
     * Puts a new name/value pair.
     * @param name the name
     * @param value the value
     */
    public void put(String name, int value) {
        if (!entries.containsKey(name)) {
            this.order.add(name);
        }
        this.entries.put(name, Integer.valueOf(value));
    }

    /**
     * Returns the value given a name.
     * @param name the name of the value
     * @return the value or null, if there's no value with the given name.
     */
    public Object get(String name) {
        return this.entries.get(name);
    }

    /** {@inheritDoc} */
    @Override
    public int output(OutputStream stream) throws IOException {
        CountingOutputStream cout = new CountingOutputStream(stream);
        StringBuilder textBuffer = new StringBuilder(64);
        writeDictionary(cout, textBuffer);
        PDFDocument.flushTextBuffer(textBuffer, cout);
        return cout.getCount();
    }

    /**
     * Writes the contents of the dictionary to a StringBuffer.
     * @param out the OutputStream (for binary content)
     * @param textBuffer the text buffer for text output
     * @throws IOException if an I/O error occurs
     */
    protected void writeDictionary(OutputStream out, StringBuilder textBuffer) throws IOException {
        textBuffer.append("<<");
        boolean compact = (this.order.size() <= 2);
        for (String key : this.order) {
            if (compact) {
                textBuffer.append(' ');
            } else {
                textBuffer.append("\n  ");
            }
            textBuffer.append(PDFName.escapeName(key));
            textBuffer.append(' ');
            Object obj = this.entries.get(key);
            formatObject(obj, out, textBuffer);
        }
        if (compact) {
            textBuffer.append(' ');
        } else {
            textBuffer.append('\n');
        }
        textBuffer.append(">>");
    }

    @Override
    public void getChildren(Set<PDFObject> children) {
        if (!visited) {
            visited = true;
            Map<String, Object> childrenMap = new HashMap<String, Object>(entries);
            childrenMap.remove("Parent");
            getChildren(childrenMap.values(), children);
            visited = false;
        }
    }

    public static void getChildren(Collection<Object> values, Set<PDFObject> children) {
        for (Object x : values) {
            if (x instanceof PDFReference) {
                x = ((PDFReference) x).getObject();
            }
            if (x instanceof PDFObject) {
                if (((PDFObject) x).hasObjectNumber()) {
                    children.add((PDFObject) x);
                }
                ((PDFObject) x).getChildren(children);
            }
        }
    }

    public Set<String> keySet() {
        return entries.keySet();
    }

    /**
     * @see java.util.Map#containsKey(Object)
     *
     * @param name The key to find in the map.
     * @return true if the map contains this key.
     */
    public boolean containsKey(String name) {
        return this.entries.containsKey(name);
    }

    /**
     * Removes the mapping for the specified key
     * @param name  key whose mapping is to be removed
     */
    public void remove(String name) {
        entries.remove(name);
    }
}

File: dotnet/src/Fop.Pdf/PDFDictionary.cs
using System.Text;

namespace Fop.Pdf;

/// <summary>
/// A PDF dictionary object.
/// </summary>
public class PDFDictionary : PDFObject
{
    private bool _visited;

    /// <summary>The entry map.</summary>
    protected Dictionary<string, object?> Entries { get; } = [];

    /// <summary>
    /// Maintains the insertion order of the entries. Kept in sync with <see cref="Entries"/>.
    /// </summary>
    protected List<string> Order { get; } = [];

    /// <summary>Creates a new dictionary object.</summary>
    public PDFDictionary()
    {
    }

    /// <summary>Creates a new dictionary object.</summary>
    /// <param name="parent">The object's parent, if any.</param>
    public PDFDictionary(PDFObject? parent) : base(parent)
    {
    }

    /// <summary>Puts a new name/value pair.</summary>
    public void Put(string name, object? value)
    {
        if (value is PDFObject pdfObj && !pdfObj.HasObjectNumber)
        {
            pdfObj.Parent = this;
        }
        if (!Entries.ContainsKey(name))
        {
            Order.Add(name);
        }
        Entries[name] = value;
    }

    /// <summary>Puts a new name/value pair with an int value.</summary>
    public void Put(string name, int value)
    {
        if (!Entries.ContainsKey(name))
        {
            Order.Add(name);
        }
        Entries[name] = value;
    }

    /// <summary>Returns the value for a name, or <see langword="null"/> if absent.</summary>
    public object? Get(string name) => Entries.GetValueOrDefault(name);

    /// <summary>Gets or sets the value for a name. The getter returns <see langword="null"/> if absent.</summary>
    public object? this[string name]
    {
        get => Get(name);
        set => Put(name, value);
    }

    /// <inheritdoc/>
    public override int Output(Stream stream)
    {
        var cout = new CountingStream(stream);
        var textBuffer = new StringBuilder(64);
        WriteDictionary(cout, textBuffer);
        FlushTextBuffer(textBuffer, cout);
        return (int)cout.Count;
    }

    /// <summary>Writes the contents of the dictionary to the text buffer (and stream for binary).</summary>
    protected void WriteDictionary(Stream output, StringBuilder textBuffer)
    {
        textBuffer.Append("<<");
        bool compact = Order.Count <= 2;
        foreach (string key in Order)
        {
            textBuffer.Append(compact ? " " : "\n  ");
            textBuffer.Append(PDFName.EscapeName(key));
            textBuffer.Append(' ');
            // Use GetValueOrDefault so a key left in Order after Remove() (matching Java's
            // entries/order desync) serializes as "null" rather than throwing.
            FormatObject(Entries.GetValueOrDefault(key), output, textBuffer);
        }
        textBuffer.Append(compact ? " " : "\n");
        textBuffer.Append(">>");
    }

    /// <inheritdoc/>
    public override void GetChildren(ISet<PDFObject> children)
    {
        if (!_visited)
        {
            _visited = true;
            var childrenMap = new Dictionary<string, object?>(Entries);
            childrenMap.Remove("Parent");
            GetChildren(childrenMap.Values, children);
            _visited = false;
        }
    }

    /// <summary>Collects the descendant PDF objects (with assigned object numbers) of the given values.</summary>
    public static void GetChildren(IEnumerable<object?> values, ISet<PDFObject> children)
    {
        foreach (object? value in values)
        {
            object? x = value;
            if (x is PDFReference reference)
            {
                x = reference.Object;
            }
            if (x is PDFObject pdfObject)
            {
                if (pdfObject.HasObjectNumber)
                {
                    children.Add(pdfObject);
                }
                pdfObject.GetChildren(children);
            }
        }
    }

    /// <summary>The set of keys in this dictionary.</summary>
    public ICollection<string> Keys => Entries.Keys;

    /// <summary>Indicates whether the dictionary contains the given key.</summary>
    public bool ContainsKey(string name) => Entries.ContainsKey(name);

    /// <summary>Removes the mapping for the specified key.</summary>
    public void Remove(string name) => Entries.Remove(name);
}
